### DISCUSSION TITLE
flax-specific tools for diagramming a flax nn?

### DISCUSSION THREAD

**Discussion by grahamgower** [google/flax#2032] [Q&A]
There's a nice list of diagramming tools for neural networks [here]( https://github.com/ashishpatel26/Tools-to-Design-or-Visualize-Architecture-of-Neural-Network), some of which work directly off models implemented for keras. Are there any such tools that do this for flax networks?

**Comment by marcvanzee** [comment]
Hi @grahamgower, you could convert you Flax module to TF using [jax2tf](https://github.com/google/jax/tree/main/jax/experimental/jax2tf) and then use Tensorflow's diagramming tools in Tensorboard to visualize your model.

**Reply by grahamgower** [reply]
Thanks for the suggestion, but I find tensorboard's visuals to be more confusing than illuminating.

**Reply by marcvanzee** [reply]
Yes, they can be quite difficult to grasp, I agree. Another thing you could try is @cgarciae's recently introduced [flax.linen.tabulate](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.tabulate.html), which outputs Modules with various useful statistics in a tabular way, where you can select the depth of the submodule tree manually.

**Reply by grahamgower** [reply]
Ah, yes, tabulate is awesome and I'm already using it! And once https://github.com/google/jax/issues/4085 is fixed, then it will be even better.

**Reply by marcvanzee** [reply]
Even better, when #2316 is in the submodules will be sorted by call order 😃  (it should be in one of these days).
